File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/AssetUploadRequest.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.nio.ByteBuffer;
import java.util.UUID;

/**
 * Asset storage messages
 * current assumes an existing UUID, need to enhance for new assets
 *
 * <p>Template: {@code AssetUploadRequest Low 333 NotTrusted Unencoded}
 * (recovered/reference/message_template.msg).
 */
public class AssetUploadRequest extends SLMessage {
    public AssetBlock AssetBlock_Field;

    /** Block AssetBlock, Single. */
    public static class AssetBlock {
        public byte[] AssetData; // Variable 2 - Optional: the actual asset data if the whole thing will fit it this packet
        public boolean StoreLocal; // BOOL
        public boolean Tempfile; // BOOL
        public UUID TransactionID; // LLUUID
        public int Type; // S8
    }

    public AssetUploadRequest() {
        this.zeroCoded = false;
        this.AssetBlock_Field = new AssetBlock();
    }

    @Override
    public int CalcPayloadSize() {
        return this.AssetBlock_Field.AssetData.length + 21 + 4;
    }

    @Override
    public void Handle(SLMessageHandler sLMessageHandler) {
        sLMessageHandler.HandleAssetUploadRequest(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: Low 333 (AssetUploadRequest).
        byteBuffer.putShort((short) 0xFFFF);
        byteBuffer.put((byte) 0x01);
        byteBuffer.put((byte) 0x4D);
        packUUID(byteBuffer, this.AssetBlock_Field.TransactionID);
        packByte(byteBuffer, (byte) this.AssetBlock_Field.Type);
        packBoolean(byteBuffer, this.AssetBlock_Field.Tempfile);
        packBoolean(byteBuffer, this.AssetBlock_Field.StoreLocal);
        packVariable(byteBuffer, this.AssetBlock_Field.AssetData, 2);
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.AssetBlock_Field.TransactionID = unpackUUID(byteBuffer);
        this.AssetBlock_Field.Type = unpackByte(byteBuffer);
        this.AssetBlock_Field.Tempfile = unpackBoolean(byteBuffer);
        this.AssetBlock_Field.StoreLocal = unpackBoolean(byteBuffer);
        this.AssetBlock_Field.AssetData = unpackVariable(byteBuffer, 2);
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/AssetUploadRequest.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import java.nio.ByteBuffer
import java.util.UUID

/**
 * Asset storage messages
 * current assumes an existing UUID, need to enhance for new assets
 *
 * <p>Template: {@code AssetUploadRequest Low 333 NotTrusted Unencoded}
 * (recovered/reference/message_template.msg).
 */
open class AssetUploadRequest : SLMessage() {
    @JvmField var AssetBlock_Field: AssetBlock = AssetBlock()

    /** Block AssetBlock, Single. */
    open class AssetBlock {
        @JvmField var AssetData: ByteArray? = null
        @JvmField var StoreLocal: Boolean = false
        @JvmField var Tempfile: Boolean = false
        @JvmField var TransactionID: UUID? = null
        @JvmField var Type: Int = 0
    }

    init {
        zeroCoded = false
    }

    override fun CalcPayloadSize(): Int {
        return AssetBlock_Field.AssetData!!.size + 21 + 4
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandleAssetUploadRequest(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: Low 333 (AssetUploadRequest).
        byteBuffer.putShort(0xFFFF.toShort())
        byteBuffer.put((0x01).toByte())
        byteBuffer.put((0x4D).toByte())
        packUUID(byteBuffer, AssetBlock_Field.TransactionID)
        packByte(byteBuffer, (AssetBlock_Field.Type).toByte())
        packBoolean(byteBuffer, AssetBlock_Field.Tempfile)
        packBoolean(byteBuffer, AssetBlock_Field.StoreLocal)
        packVariable(byteBuffer, AssetBlock_Field.AssetData, 2)
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        AssetBlock_Field.TransactionID = unpackUUID(byteBuffer)
        AssetBlock_Field.Type = unpackByte(byteBuffer).toInt()
        AssetBlock_Field.Tempfile = unpackBoolean(byteBuffer)
        AssetBlock_Field.StoreLocal = unpackBoolean(byteBuffer)
        AssetBlock_Field.AssetData = unpackVariable(byteBuffer, 2)
    }
}
